File: Week_10/Practicum/Count the Number of Complete Components.py
def dfs(graph, visited, current, current_component):
    if visited[current]:
        return

    current_component.append(current)

    visited[current] = True

    for node in graph[current]:
        dfs(graph, visited, node, current_component)

class Solution:
    def countCompleteComponents(self, n: int, edges: List[List[int]]) -> int:
        graph = [[] for _ in range(n)]
        for edge in edges:
            graph[edge[0]].append(edge[1])
            graph[edge[1]].append(edge[0])

        visited = [False for node in range(n)]

        connected_components = []

        for node in range(n):
            if not visited[node]:
                current_component = []
                dfs(graph, visited, node, current_component)
                connected_components.append(current_component)

        count = 0

        for component in connected_components:
            complete = True
            for node in component:
                if len(graph[node]) != len(component) - 1:
                    complete = False
                    break
            count = count + 1 if complete else count

        return count

File: Week_10/Practicum/Count the Number of Complete Components.cpp
class Solution {
public:
    void dfs(vector<vector<int>>& graph, vector<bool>& visited, int curr, int& vertices, int& edges) {
        visited[curr] = true;
        edges += graph[curr].size();
        vertices++;

        for (auto& n : graph[curr]) {
            if (!visited[n]) {
                dfs(graph, visited, n, vertices, edges);
            }
        }
    }

    int countCompleteComponents(int n, vector<vector<int>>& edges) {
        vector<vector<int>> graph(n, vector<int>());

        for (auto& edge : edges) {
            graph[edge[0]].push_back(edge[1]);
            graph[edge[1]].push_back(edge[0]);
        }

        vector<bool> visited(n, false);
        int counter = 0;
        for (int i = 0; i < n; i++) {
            if (!visited[i]) {
                int vertices = 0;
                int edges = 0;
                dfs(graph,visited, i, vertices, edges);

                if (vertices * (vertices - 1) == edges) {
                    counter++;
                }
            }
        }
        return counter;
    }
};